apps/builder/app/builder/shared/assets/use-assets.tsx

```diff
@@ -108,22 +106,63 @@ const assetContainersStore = computed(
 
 export type UploadData = ActionData;
 
-export const useUploadAsset = () => {
-  const submitAsset = usePersistentFetcher();
-  const submitUpload = usePersistentFetcher();
-
-  const handleAfterSubmit = (assetId: string, data: UploadData) => {
-    // remove uploaded or failed asset
-    deleteUploadingFileData(assetId);
+const uploadAsset = async ({
+  authToken,
+  projectId,
+  assetId,
+  type,
+  file,
+  onCompleted,
+  onError,
+}: {
+  authToken: undefined | string;
+  projectId: string;
+  assetId: string;
+  type: string;
+  file: File;
+  onCompleted: (data: UploadData) => void;
+  onError: (error: string) => void;
+}) => {
+  try {
+    const metaFormData = new FormData();
+    metaFormData.append("projectId", projectId);
+    metaFormData.append("assetId", assetId);
+    // sanitizeS3Key here is just because of https://github.com/remix-run/remix/issues/4443
+    // should be removed after fix
+    metaFormData.append("filename", sanitizeS3Key(file.name));
+    const metaResponse = await fetch(restAssetsPath({ authToken }), {
+      method: "POST",
+      body: metaFormData,
+    });
+    const metaData: { name: string } | { errors: string } =
+      await metaResponse.json();
+    if ("errors" in metaData) {
+      throw Error(metaData.errors);
```

Comment:
`new Error`?

Reply:
Never used new for old school builtin types :)